grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcHealthCheckService.java

```diff
@@ -0,0 +1,279 @@
+ * @see GrpcHealthCheckServiceBuilder
+ */
+@UnstableApi
+public final class GrpcHealthCheckService extends HealthImplBase {
+
+    private static final Logger logger = LoggerFactory.getLogger(GrpcHealthCheckService.class);
+
+    private static final HealthCheckResponse SERVING_RESPONSE =
+            HealthCheckResponse.newBuilder()
+                               .setStatus(ServingStatus.SERVING)
+                               .build();
+
+    private static final HealthCheckResponse NOT_SERVING_RESPONSE =
+            HealthCheckResponse.newBuilder()
+                               .setStatus(ServingStatus.NOT_SERVING)
+                               .build();
+
+    private static final HealthCheckResponse SERVICE_UNKNOWN_RESPONSE =
+            HealthCheckResponse.newBuilder()
+                               .setStatus(ServingStatus.SERVICE_UNKNOWN)
+                               .build();
+
+    private static final String EMPTY_SERVICE = "";
+
+    /**
+     * Returns a newly created {@link GrpcHealthCheckService}
+     * with the specified {@link ListenableHealthChecker}s.
+     */
+    public static GrpcHealthCheckService of(ListenableHealthChecker... healthCheckers) {
+        return builder().checkers(healthCheckers).build();
+    }
+
+    /**
+     * Returns a newly created {@link GrpcHealthCheckService}
+     * with the specified {@link ListenableHealthChecker}s.
+     */
+    public static GrpcHealthCheckService of(Iterable<? extends ListenableHealthChecker> healthCheckers) {
+        return builder().checkers(healthCheckers).build();
+    }
+
+    /**
+     * Returns a new builder which builds a new {@link GrpcHealthCheckService}.
+     */
+    public static GrpcHealthCheckServiceBuilder builder() {
+        return new GrpcHealthCheckServiceBuilder();
+    }
+
+    private static HealthCheckResponse getHealthCheckResponse(ServingStatus status) {
+        switch (status) {
+            case SERVING:
+                return SERVING_RESPONSE;
+            case NOT_SERVING:
+                return NOT_SERVING_RESPONSE;
+            case SERVICE_UNKNOWN:
+                return SERVICE_UNKNOWN_RESPONSE;
+            default:
+                throw new IllegalArgumentException("Invalid status:" + status);
+        }
+    }
+
+    private static StatusRuntimeException getNotFoundStatus(String service) {
+        return Status.NOT_FOUND
+                .withDescription(String.format("The service name(%s) is not registered in this service",
+                                               service))
+                .asRuntimeException();
+    }
+
+    private final SettableHealthChecker serverHealth;
+    private final Set<ListenableHealthChecker> serverHealthCheckers;
+    private final Map<String, ListenableHealthChecker> grpcServiceHealthCheckers;
+    private final Multimap<String, StreamObserver<HealthCheckResponse>> watchers = HashMultimap.create();
```

Comment:
not strong on this, but wondering if using a `IdentityHashSet` is better since `StreamObserver` is a public interface

```suggestion
    private final Multimap<String, StreamObserver<HealthCheckResponse>> watchers = Multimaps.newSetMultimap(new HashMap<>(), Sets::newIdentityHashSet);
```

Reply:
I'm not sure what your concern is.
I thought the API behavior (watch method) would be the same, so I adopted it.

Reply:
The default set implementation of `HashMultimap` is `HashSet` which evaluates equality depending on how `HealthCheckResponse.equals()` is defined.

Having said this, I don't think any prominent implementations out there override `equals()`, so I guess it's fine to leave as is